src/util/tmpl/fd_heap.c

```diff
@@ -0,0 +1,647 @@
+  ulong      ele_cnt; /* Current number of elements in heap, in [0,ele_max] */
+  HEAP_IDX_T root;    /* Index of the root heap element, in [0,ele_max) or HEAP_IDX_NULL */
+};
+
+typedef struct HEAP_(private) HEAP_(t);
+
+FD_PROTOTYPES_BEGIN
+
+/* prototypes */
+
+HEAP_STATIC FD_FN_CONST ulong      HEAP_(align)    ( void                             );
+HEAP_STATIC FD_FN_CONST ulong      HEAP_(footprint)( ulong      ele_max               );
+HEAP_STATIC /**/        void *     HEAP_(new)      ( void *     shmem,  ulong ele_max );
+HEAP_STATIC /**/        HEAP_(t) * HEAP_(join)     ( void *     shheap                );
+HEAP_STATIC /**/        void *     HEAP_(leave)    ( HEAP_(t) * heap                  );
+HEAP_STATIC /**/        void *     HEAP_(delete)   ( void *     shheap                );
+
+HEAP_STATIC HEAP_(t) * HEAP_(idx_insert)( HEAP_(t) * heap, ulong n, HEAP_T * ele );
+
+HEAP_STATIC HEAP_(t) * HEAP_(idx_remove_min)( HEAP_(t) * heap, HEAP_T * ele );
+
+HEAP_STATIC FD_FN_PURE int HEAP_(verify)( HEAP_(t) const * heap, HEAP_T const * ele );
+
+/* inlines */
+
+FD_FN_PURE static inline int HEAP_(lt) ( HEAP_T const * e0, HEAP_T const * e1 ) { return HEAP_LT( e0, e1 ); }
+
+FD_FN_CONST static inline ulong          HEAP_(idx_null)      ( void ) { return HEAP_IDX_NULL; }
+FD_FN_CONST static inline HEAP_T *       HEAP_(ele_null)      ( void ) { return NULL;          }
+FD_FN_CONST static inline HEAP_T const * HEAP_(ele_null_const)( void ) { return NULL;          }
+
+FD_FN_CONST static inline int HEAP_(idx_is_null)( ulong          i ) { return HEAP_IDX_IS_NULL( i ); }
+FD_FN_CONST static inline int HEAP_(ele_is_null)( HEAP_T const * e ) { return !e;                    }
+
+FD_FN_CONST static inline ulong
+HEAP_(idx)( HEAP_T const * e,
+            HEAP_T const * ele ) {
+  return fd_ulong_if( !!e, (ulong)(e-ele), HEAP_IDX_NULL );
+}
+
+FD_FN_CONST static inline HEAP_T *
+HEAP_(ele)( ulong    i,
+            HEAP_T * ele ) {
+  return fd_ptr_if( !HEAP_IDX_IS_NULL( i ), ele + i, NULL );
+}
+
+FD_FN_CONST static inline HEAP_T const *
+HEAP_(ele_const)( ulong          i,
+                  HEAP_T const * ele ) {
+  return fd_ptr_if( !HEAP_IDX_IS_NULL( i ), ele + i, NULL );
+}
+
+FD_FN_CONST static inline ulong          HEAP_(idx_fast)      ( HEAP_T const * e, HEAP_T const * ele ) { return (ulong)(e - ele); }
+FD_FN_CONST static inline HEAP_T *       HEAP_(ele_fast)      ( ulong i,          HEAP_T *       ele ) { return ele + i; }
+FD_FN_CONST static inline HEAP_T const * HEAP_(ele_fast_const)( ulong i,          HEAP_T const * ele ) { return ele + i; }
+
+FD_FN_PURE static inline ulong HEAP_(ele_max)( HEAP_(t) const * heap ) { return heap->ele_max; }
+FD_FN_PURE static inline ulong HEAP_(ele_cnt)( HEAP_(t) const * heap ) { return heap->ele_cnt; }
+
+FD_FN_PURE static inline ulong HEAP_(idx_peek_min)( HEAP_(t) const * heap ) { return (ulong)heap->root; }
+
+FD_FN_PURE static inline HEAP_T *
+HEAP_(ele_peek_min)( HEAP_(t) const * heap,
```

Comment:
In this API and others, the meaning of the ele argument is somewhat mysterious. Is this a well known sentinel?

Reply:
See global comment below.